Products/CMFPlone/PasswordResetTool.py

```diff
@@ -0,0 +1,323 @@
+"""PasswordResetTool.py
+
+Mailback password reset product for CMF.
+Author: J Cameron Cooper, Sept 2003
+"""
+from hashlib import md5
+
+from Products.CMFCore.utils import UniqueObject
+from Products.CMFCore.utils import getToolByName
+from OFS.SimpleItem import SimpleItem
+from App.class_init import InitializeClass
+from AccessControl import ClassSecurityInfo
+from AccessControl import ModuleSecurityInfo
+from plone.uuid.interfaces import IUUIDGenerator
+from plone.registry.interfaces import IRegistry
+from Products.CMFCore.permissions import ManagePortal
+from Products.CMFPlone.interfaces import IPWResetTool
+from Products.CMFPlone.interfaces import ISecuritySchema
+from Products.CMFPlone.RegistrationTool import get_member_by_login_name
+
+import datetime
+import time
+from DateTime import DateTime
+from zope.component import getUtility
+from zope.interface import implementer
+
+module_security = ModuleSecurityInfo('Products.CMFPlone.PasswordResetTool')
+
+module_security.declarePublic('InvalidRequestError')
+class InvalidRequestError(Exception):
+    """ Request reset URL is invalid """
+    def __init__(self, value=''):
+        self.value = value
+
+    def __str__(self):
+        return repr(self.value)
+
+module_security.declarePublic('ExpiredRequestError')
+class ExpiredRequestError(InvalidRequestError):
+    """ Request reset URL is expired """
+
+
+@implementer(IPWResetTool)
+class PasswordResetTool (UniqueObject, SimpleItem):
+    """Provides a default implementation for a password reset scheme.
+
+    From a 'forgotten password' template, you submit your username to
+    a handler script that does a 'requestReset', and sends an email
+    with an unguessable unique hash in a url as built by 'constructURL'
+    to the user.
+
+    The user visits that URL (the 'reset form') and enters their username,
+    """
+
+    id = 'portal_password_reset'
+    meta_type = 'Password Reset Tool'
+
+    security = ClassSecurityInfo()
+
+    security.declareProtected(ManagePortal, 'manage_setTimeout')
+
+    def manage_setTimeout(self, hours=168, REQUEST=None):
+        """ZMI method for setting the expiration timeout in hours."""
+        self.setExpirationTimeout(int(hours))
+        return self.manage_overview(manage_tabs_message="Timeout set to %s hours" % hours)
+
+    security.declareProtected(ManagePortal, 'manage_toggleUserCheck')
+
+    def manage_toggleUserCheck(self, REQUEST=None):
+        """ZMI method for toggling the flag for checking user names on return.
+        """
+        self.toggleUserCheck()
+        m = self.checkUser() and 'on' or 'off'
+        return self.manage_overview(manage_tabs_message="Returning username check turned %s" % m)
+
+    def __init__(self):
+        self._requests = {}
```

Comment:
While we're moving the password tool, let's change this to a BTree. We had a customer with a site that had self-registration turned on and was getting spam accounts created, and the database grew rapidly because each new user saved a new copy of this entire dict.